### PR TITLE
Avro json payload

### PR DESCRIPTION
**Pull request by dswiecki**

### CODE REVIEW

File: utils/avro-components-utils/src/main/scala/pl/touk/nussknacker/engine/avro/schemaregistry/confluent/client/AvroWithJsonPayloadSchema.scala [hunk 1]
```diff
@@ -0,0 +1,22 @@
+package pl.touk.nussknacker.engine.avro.schemaregistry.confluent.client
+
+import io.confluent.kafka.schemaregistry.ParsedSchema
+import io.confluent.kafka.schemaregistry.avro.AvroSchema
+import io.confluent.kafka.schemaregistry.client.rest.entities.SchemaReference
+import org.apache.avro.Schema
+
+import java.util
+
+case class AvroWithJsonPayloadSchema(avroSchema: AvroSchema) extends ParsedSchema {
```
Comment (gadomsky) [inline utils/avro-components-utils/src/main/scala/pl/touk/nussknacker/engine/avro/schemaregistry/confluent/client/AvroWithJsonPayloadSchema.scala:10]: `AvroSchemaWithJsonPayload` ?

File: utils/avro-components-utils/src/main/scala/pl/touk/nussknacker/engine/avro/schemaregistry/confluent/client/CachedConfluentSchemaRegistryClient.scala [hunk 2]
```diff
@@ -54,9 +55,27 @@ class CachedConfluentSchemaRegistryClient(val client: CSchemaRegistryClient, cac
...
+
+  override def getSchemaById(id: Int): SchemaWithMetadata = {
+    //todo schemaCache is subject-version keyed, how to get it by id
+    val rawSchema = client.getSchemaById(id)
+    withExtraSchemaTypes(id, rawSchema)
+  }
+
+  private def withExtraSchemaTypes(id: Int, rawSchema: ParsedSchema) = {
+    (rawSchema.schemaType(), config.avroPlainTextSerialization) match {
+      case ("AVRO", Some(true)) => SchemaWithMetadata(AvroWithJsonPayloadSchema(rawSchema.asInstanceOf[AvroSchema]), id)
```
Comment (gadomsky) [inline utils/avro-components-utils/src/main/scala/pl/touk/nussknacker/engine/avro/schemaregistry/confluent/client/CachedConfluentSchemaRegistryClient.scala:70]: `case (schema: AvroSchema, Some(true) => ` then, you wil get rid of casting in the following code

File: utils/kafka-utils/src/main/scala/pl/touk/nussknacker/engine/kafka/KafkaConfig.scala [hunk 3]
```diff
@@ -7,7 +7,8 @@ import scala.concurrent.duration._
 
 case class SchemaRegistryClientKafkaConfig(
                                             kafkaProperties: Map[String, String],
-                                            cacheConfig: SchemaRegistryCacheConfig
+                                            cacheConfig: SchemaRegistryCacheConfig,
+                                            avroPlainTextSerialization: Option[Boolean]
```
Comment (mproch) [inline utils/kafka-utils/src/main/scala/pl/touk/nussknacker/engine/kafka/KafkaConfig.scala:11]: avroAsJsonSerialization? I wouldn't call json plain text ;)

File: utils/avro-components-utils/src/main/scala/pl/touk/nussknacker/engine/avro/schemaregistry/confluent/serialization/AbstractConfluentKafkaAvroDeserializer.scala [hunk 4]
```diff
@@ -65,6 +67,18 @@ class ConfluentAvroPayloadDeserializer(
   }
 }
 
+object ConfluentJsonPayloadDeserializer extends UniversalSchemaPayloadDeserializer {
+
+  override def deserialize(expectedSchemaData: Option[RuntimeSchemaData[ParsedSchema]], writerSchemaData: RuntimeSchemaData[ParsedSchema], buffer: ByteBuffer, bufferDataStart: Int): AnyRef = {
+    //todo refactor
+    val converter = new JsonPayloadToAvroConverter(None)
```
Comment (mproch) [inline utils/avro-components-utils/src/main/scala/pl/touk/nussknacker/engine/avro/schemaregistry/confluent/serialization/AbstractConfluentKafkaAvroDeserializer.scala:74]: shouldn't be instance field?

File: utils/avro-components-utils/src/main/scala/pl/touk/nussknacker/engine/avro/schemaregistry/confluent/serialization/AbstractConfluentKafkaAvroDeserializer.scala [hunk 5]
```diff
@@ -65,6 +67,18 @@ class ConfluentAvroPayloadDeserializer(
   }
 }
 
+object ConfluentJsonPayloadDeserializer extends UniversalSchemaPayloadDeserializer {
+
+  override def deserialize(expectedSchemaData: Option[RuntimeSchemaData[ParsedSchema]], writerSchemaData: RuntimeSchemaData[ParsedSchema], buffer: ByteBuffer, bufferDataStart: Int): AnyRef = {
+    //todo refactor
```
Comment (mproch) [inline utils/avro-components-utils/src/main/scala/pl/touk/nussknacker/engine/avro/schemaregistry/confluent/serialization/AbstractConfluentKafkaAvroDeserializer.scala:73]: what should be refactored?

File: utils/avro-components-utils/src/main/scala/pl/touk/nussknacker/engine/avro/schemaregistry/confluent/client/ConfluentSchemaRegistryClient.scala [hunk 6]
```diff
@@ -59,6 +62,22 @@ class DefaultConfluentSchemaRegistryClient(override val client: CSchemaRegistryC
       client.getAllVersions(subject).asScala.toList
     }
 
+  override def getSchemaById(id: Int): SchemaWithMetadata = {
+    //todo schemaCache is subject-version keyed, how to get it by id
```
Comment (mproch) [inline utils/avro-components-utils/src/main/scala/pl/touk/nussknacker/engine/avro/schemaregistry/confluent/client/ConfluentSchemaRegistryClient.scala:66]: what do we want to do here?

File: utils/avro-components-utils/src/main/scala/pl/touk/nussknacker/engine/avro/schemaregistry/confluent/client/ConfluentSchemaRegistryClient.scala [hunk 7]
```diff
@@ -59,6 +62,22 @@ class DefaultConfluentSchemaRegistryClient(override val client: CSchemaRegistryC
...
+    val rawSchema = client.getSchemaById(id)
+    withExtraSchemaTypes(id, rawSchema)
+  }
+
+  private def withExtraSchemaTypes(id: Int, rawSchema: ParsedSchema) = {
```
Comment (mproch) [inline utils/avro-components-utils/src/main/scala/pl/touk/nussknacker/engine/avro/schemaregistry/confluent/client/ConfluentSchemaRegistryClient.scala:71]: I wonder, maybe let's put it in SchemaWithMetadata.apply? So we have schema dispatch logic in one pllace?

File: utils/avro-components-utils/src/main/scala/pl/touk/nussknacker/engine/avro/schemaregistry/SchemaRegistryClient.scala [hunk 8]
```diff
@@ -4,6 +4,8 @@ import cats.data.Validated
 
 trait SchemaRegistryClient extends Serializable {
 
+  def getSchemaById(id: Int): SchemaWithMetadata
```
Comment (mproch) [inline utils/avro-components-utils/src/main/scala/pl/touk/nussknacker/engine/avro/schemaregistry/SchemaRegistryClient.scala:7]: I think we don't use it?

**Review by mproch — COMMENTED** [review]
Looks pretty good, added some minor comments

File: utils/avro-components-utils/src/main/scala/pl/touk/nussknacker/engine/avro/schemaregistry/confluent/client/ConfluentSchemaRegistryClient.scala [hunk 9]
```diff
@@ -3,11 +3,14 @@ package pl.touk.nussknacker.engine.avro.schemaregistry.confluent.client
...
 import io.confluent.kafka.schemaregistry.client.rest.exceptions.RestClientException
 import io.confluent.kafka.schemaregistry.client.{SchemaRegistryClient => CSchemaRegistryClient}
-import pl.touk.nussknacker.engine.avro.AvroUtils
 import pl.touk.nussknacker.engine.avro.schemaregistry._
 import pl.touk.nussknacker.engine.avro.schemaregistry.confluent.ConfluentUtils
+import pl.touk.nussknacker.engine.kafka.SchemaRegistryClientKafkaConfig
+
 import scala.collection.JavaConverters._
 
 trait ConfluentSchemaRegistryClient extends SchemaRegistryClient with LazyLogging {
```
Comment (gadomsky) [inline utils/avro-components-utils/src/main/scala/pl/touk/nussknacker/engine/avro/schemaregistry/confluent/client/ConfluentSchemaRegistryClient.scala:16]: can we make method
```
def client: CSchemaRegistryClient
```
protected?
or enforce in other way, that noone directly calls original client? (if not, maybe we should implement delegate to `io.confluent.kafka.schemaregistry.client.SchemaRegistryClient` and map to new schema type there)?

File: utils/kafka-utils/src/main/scala/pl/touk/nussknacker/engine/kafka/KafkaConfig.scala (hunk 3 above)
Comment (gadomsky) [reply utils/kafka-utils/src/main/scala/pl/touk/nussknacker/engine/kafka/KafkaConfig.scala:11]: `avroSchemaAsJsonSerialization`?

File: utils/avro-components-utils/src/main/scala/pl/touk/nussknacker/engine/avro/schemaregistry/SchemaRegistryClient.scala (hunk 8 above)
Comment (dswiecki) [reply utils/avro-components-utils/src/main/scala/pl/touk/nussknacker/engine/avro/schemaregistry/SchemaRegistryClient.scala:7]: we use it in ConfluentUniversalKafkaDeserializer

File: utils/avro-components-utils/src/main/scala/pl/touk/nussknacker/engine/avro/schemaregistry/confluent/client/ConfluentSchemaRegistryClient.scala (hunk 9 above)
Comment (dswiecki) [reply utils/avro-components-utils/src/main/scala/pl/touk/nussknacker/engine/avro/schemaregistry/confluent/client/ConfluentSchemaRegistryClient.scala:16]: Let's do it in separate PR

File: utils/avro-components-utils/src/main/scala/pl/touk/nussknacker/engine/avro/schemaregistry/confluent/client/ConfluentSchemaRegistryClient.scala (hunk 7 above)
Comment (dswiecki) [reply utils/avro-components-utils/src/main/scala/pl/touk/nussknacker/engine/avro/schemaregistry/confluent/client/ConfluentSchemaRegistryClient.scala:71]: done

File: utils/avro-components-utils/src/main/scala/pl/touk/nussknacker/engine/avro/schemaregistry/confluent/client/ConfluentSchemaRegistryClient.scala (hunk 6 above)
Comment (dswiecki) [reply utils/avro-components-utils/src/main/scala/pl/touk/nussknacker/engine/avro/schemaregistry/confluent/client/ConfluentSchemaRegistryClient.scala:66]: it was meant for cached version - it's not obvious how to cache this by id - but for now I removed todo

File: utils/avro-components-utils/src/main/scala/pl/touk/nussknacker/engine/avro/schemaregistry/confluent/serialization/AbstractConfluentKafkaAvroDeserializer.scala (hunk 5 above)
Comment (dswiecki) [reply utils/avro-components-utils/src/main/scala/pl/touk/nussknacker/engine/avro/schemaregistry/confluent/serialization/AbstractConfluentKafkaAvroDeserializer.scala:73]: remove compact - done

File: utils/avro-components-utils/src/main/scala/pl/touk/nussknacker/engine/avro/schemaregistry/confluent/serialization/AbstractConfluentKafkaAvroDeserializer.scala (hunk 4 above)
Comment (dswiecki) [reply utils/avro-components-utils/src/main/scala/pl/touk/nussknacker/engine/avro/schemaregistry/confluent/serialization/AbstractConfluentKafkaAvroDeserializer.scala:74]: done

File: utils/avro-components-utils/src/main/scala/pl/touk/nussknacker/engine/avro/schemaregistry/SchemaRegistryClient.scala (hunk 8 above)
Comment (mproch) [reply utils/avro-components-utils/src/main/scala/pl/touk/nussknacker/engine/avro/schemaregistry/SchemaRegistryClient.scala:7]: true, had it hidden :)

File: utils/kafka-utils/src/main/scala/pl/touk/nussknacker/engine/kafka/KafkaConfig.scala [hunk 10]
```diff
@@ -23,15 +24,16 @@ case class KafkaConfig(kafkaAddress: String,
                        // when complex key with its own schema is provided, this flag is false
                        // and all topics related to this config require both key and value schema definitions.
                        useStringForKey: Boolean = true,
-                       schemaRegistryCacheConfig: SchemaRegistryCacheConfig = SchemaRegistryCacheConfig()
+                       schemaRegistryCacheConfig: SchemaRegistryCacheConfig = SchemaRegistryCacheConfig(),
+                       avroPlainTextSerialization: Option[Boolean] = None
```
Comment (mproch) [inline utils/kafka-utils/src/main/scala/pl/touk/nussknacker/engine/kafka/KafkaConfig.scala:28]: why do we need it? As we already have avroAsJsonSerialization?

File: utils/avro-components-utils/src/main/scala/pl/touk/nussknacker/engine/avro/schemaregistry/confluent/serialization/AbstractConfluentKafkaAvroDeserializer.scala [hunk 11]
```diff
@@ -65,6 +67,19 @@ class ConfluentAvroPayloadDeserializer(
   }
 }
 
+object ConfluentJsonPayloadDeserializer extends UniversalSchemaPayloadDeserializer {
+
+  val converter = new JsonPayloadToAvroConverter(None)
```
Comment (mproch) [inline utils/avro-components-utils/src/main/scala/pl/touk/nussknacker/engine/avro/schemaregistry/confluent/serialization/AbstractConfluentKafkaAvroDeserializer.scala:72]: private ;)

Comment (dswiecki) [reply utils/avro-components-utils/src/main/scala/pl/touk/nussknacker/engine/avro/schemaregistry/confluent/serialization/AbstractConfluentKafkaAvroDeserializer.scala]: done

File: utils/kafka-utils/src/main/scala/pl/touk/nussknacker/engine/kafka/KafkaConfig.scala (hunk 10 above)
Comment (dswiecki) [reply utils/kafka-utils/src/main/scala/pl/touk/nussknacker/engine/kafka/KafkaConfig.scala]: forgot about second place, it comes from kafkaConfig and we pass it SchemaRegistryClientKafkaConfig

**Review by mproch — APPROVED** [review]

**Pull request merged**